### DISCUSSION TITLE
[help] Database pulls and target rendering

### DISCUSSION THREAD

**Discussion by erinsteiner-NOAA** [ropensci/targets#1107] [Help] [answered]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I was wondering if someone could point me to the "rules" targets uses for determining whether a target that includes a database pull is rerun? I think from trial and error I've determined that if the tar_cue is set to the default and there's a database pull, the target will run every time? Is that correct?

I'd also be curious if folks have best practices they use for managing the database pulls. Right now I have a flag that sets the tar_cues for the database pull targets that I toggle on and off depending on whether I need a fresh pull. It feels clunky, does anyone have better ideas?

**Comment by wlandau** [comment, marked as answer]
`targets` only pays attention to the kinds of cues you found documented in `tar_cue()`, including changes in the return value or upstream dependencies. It is not intrinsically aware of database pulls, and in general databases are can be tricky to integrate with `targets` in a way that achieves reasonable cache invalidation behavior (c.f. https://github.com/search?q=repo%3Aropensci%2Ftargets+database&type=issues and https://github.com/search?q=repo%3Aropensci%2Ftargets+database&type=discussions). If your target tracks the upstream database table as a file and modifies the same table it pulled input from (a problematically circular workflow) then it will rerun on every `tar_make()`. On the other hand, if the target does not make any special attempt to track the database, it may never rerun even if the database table changes. It depends on your specific situation. I recommend using something like [`tarchetypes::tar_cue_force()`](https://docs.ropensci.org/tarchetypes/reference/tar_cue_force.html) or custom cue-modifying code to check the database and make a nuanced decision about whether to rerun the target.

**Reply by wlandau** [reply]
> Right now I have a flag that sets the tar_cues for the database pull targets that I toggle on and off depending on whether I need a fresh pull.

There may be ways to automate that, i.e. `tarchetypes::tar_cue_force()`, `tarchetypes::tar_change()`, or even [`tarchetypes::tar_cue_age()`](https://docs.ropensci.org/tarchetypes/reference/tar_cue_age.html)/`tar_age()` if you just want to refresh the data based on when you last pulled it.
